[DOC-2187] feat(backup): Add document for Point-in-Time Restore; (#471)
tigergraph/server-docs: modules/backup-and-restore/pages/point-in-time-restore.adoc

= Point-in-Time Restore

> The introduction of Point-in-Time Restore in version 4.1.0 is designed to enable users to restore the database to a previous time point, they can provide a timestamp and restore to it using previous made backups, even though no backup has been conducted at exactly that time point.

Reviewer: Let's improve this introduction for clarity purpose:
Point-in-Time Restore, introduced in version 4.1.0, allows users to restore the database to a specific previous time point. By providing a timestamp, users can restore the database using previous backups, even if no backup was created exactly at that time.

Author: Done

> The introduction of Point-in-Time Restore in version 4.1.0 is designed to enable users to restore the database to a previous time point, they can provide a timestamp and restore to it using previous made backups, even though no backup has been conducted at exactly that time point.

Reviewer 2: ```suggestion
Point-in-Time Restore, introduced in version 4.1.0, enables users to restore the database to a previous time point using previously made backups, even though no backup was conducted at exactly that time point.
```

This feature is based on xref:tigergraph-server:backup-and-restore:differential-backups.adoc[Differential Backup].

> == Usage

Reviewer: @liangdeLitg  Include a "Prerequisites" section to specify the necessary conditions if there are any. What about required privileges?

Author: I think it has no difference to what backup/restore needs. @runlilong Do you have more info?

Reviewer 3: @liangdeLitg  yes, there is no difference to what /backup/restore needs
cc @arun-TG-PM

> To perform point-in-time restore, run the following command:

Reviewer 2: ```suggestion
To perform point-in-time restore, run the following command.
The timepoint should be in timestamp format.
```

[console, gsql]
----
$ gadmin backup restore <timepoint_to_restore> --time-point
----

For example, if you want to restore the cluster to the timepoint `2024-06-01T14:24:31Z` , run `gadmin backup restore 2024-06-01T14:24:31Z --time-point`. remember, the timepoint is precise to only seconds, other minor time parts will be truncated.

> The point-in-time restore is based on full backup and differential backup. So we need to ensure the time point to restore is not under these cases:

Reviewer: @liangdeLitg  
So for clarity, let's rephrase these important points:

Point-in-Time Restore relies on both full and differential backups to restore the database to a specific time. When selecting a time point for restoration, it is crucial to ensure that the chosen time point meets certain conditions. Specifically, the time point must not fall under the following cases:

Time Points Before the First Full Backup or After the Current Time:
 - The selected time point cannot be earlier than the first full backup or later than the current system time. This is because the restore process needs a reference point from an existing full backup, and restoring to a future time is not possible.
 - Example: If the first full backup was made on 2024-06-01T00:00:00Z, you cannot restore to any time before this, such as 2024-05-31T23:59:59Z. Similarly, if the current time is 2024-06-15T12:00:00Z, you cannot restore to a time like 2024-06-16T00:00:00Z.

Similarly , explain clearly for the second point. It is not very clear.

Reviewer 2: I don't understand the 2nd point either. I think it has to do with the situation where you have already done a point-in-time restore in the past?  A timeline figure would help.

Author: Done

---

here, the time range for backup `backup-2024-06-13T061939.131` is `[2024-06-13T06:19:39Z,2024-06-13T06:19:39Z]`, therefore, for a full backup is just a point. As for the incremental backup `backup-2024-06-13T063915.054`, its range is `[2024-06-13T06:19:39Z , 2024-06-13T06:39:15Z]`. so, for instance, if we have performed a restore at `2024-06-14T06:19:39Z`. And then we can’t restore to the timepoint `2024-06-13T07:00:00Z`, for it is not in any backup's time range. however, `2024-06-13T06:35:00Z` is supported for it is covered by the time range `[2024-06-13T06:19:39Z , 2024-06-13T06:39:15Z]` of the backup `backup-2024-06-13T063915.054`.

[NOTE]
====
because the restore action will make the uncovered timepoint unavailable to restore, if you are not very sure about the time point to restore,  we’d better ensure the possible time points are all covered by existing backups. if there is not this kind of backup, you may create a by `gadmin backup create <backup_tag> --incremental --base <full_backup_tag>`, the new backup will cover the range from the full backup <full_backup_tag>'s create time to the current time.
====

== Key Benefits

> The primary advantage of Point-in-Time Restore is the flexibility it offers in selecting any specific time point on the timeline for restoration.

Reviewer: Let's expand the key benefits:
The primary advantage of Point-in-Time Restore is the flexibility it offers, allowing users to restore the database to a specific time point. This feature is particularly useful in scenarios where unwanted changes or data corruption need to be undone without losing other valuable work done after the last backup.

.Key Benefit Scenario
====
. A user creates a full backup on __Sunday__.

. During the week, user creates a differential backup on every day 10pm.

. On Thursday, user finds there are unwanted changes made on Wednesday around 2pm.

. User can restore system to 1:55pm of Wednesday instead of the backup made on Tuesday 10pm.

. The changes made between Tuesday 10pm and Wednesday 1:55pm are saved this way.
====

Summery of key benefits:

* Precise recover time point to save a much as possible correct work made to database.

== Limitations and Considerations

Currently, the Point-in-Time Restore feature has these known limitations:

> .Point-in-Time Restore examples.

Reviewer 2: We don't make any comment here (or for regular Restore) about how long Restore might take.  
* I assume the user should stop all other activity, correct?  What will happen if a user tries to run a command or submit a request?
* How will the database tell you that it is finished restoring?

Author: For requests like RESTPP request or query, they will fail, as starting GPE/GSE is the last step of restore, TG cannot serve those requests during restore. User will know restoring finished through information printed out at console.

@runlilong Please help answer what will happen if run gadmin command while restoring. And any other ways to know restore is finished?

Reviewer 3: > Please help answer what will happen if run gadmin command while restoring.

we assume there wouldn't be significant gadmin command during the restore.

as for the consequence is dependent on the concrete command and the fire-out time. if the command has nothing to do with restore flow and its dependent services are online,  the command can respond normally, such as `gadmin status`. however, it would fail if the restore flow is exactly at the gap of stop Kafka.  and if the gadmin command is to stop ctrl, then the restore flow will be failed.  

> And any other ways to know restore is finished?

the restore command completion means the restore is done.

@liangdeLitg  cc @victorleeTG

Reviewer 2: I won't add anything to the doc right now, but we should ask the TSE team what advice they would give to users and then add that.

image::PIT_restore.png[]

> 

Reviewer: Let's have a sub-heading for each limitations:
1. GSQL Content: The GSQL content, such as schemas and RBAC settings, cannot be restored to the exact version at the provided time point. After a Point-in-Time Restore, these elements will reflect the version of the nearest differential backup used.
2. Precision Window: Point-in-Time Restore operates with a precision window of a few minutes. To ensure optimal data consistency, it is recommended to restore to a time point at least 2-3 minutes before the timestamp of any unwanted changes.
3. Coverage: Point-in-Time Restore cannot be performed for timestamps not covered by an existing differential backup. If the desired time point is not covered, a new differential backup must be created before performing the restore."

Author: Done

= Point-in-Time Restore

Point-in-Time Restore, introduced in version 4.1.0, enables users to restore the database to a previous time point using previously made backups, even though no backup was conducted at exactly that time point.
This feature is based on xref:tigergraph-server:backup-and-restore:differential-backups.adoc[Differential Backup].

== Usage

To perform point-in-time restore, run the following command.
The timepoint should be in timestamp format.

[console, gsql]
----
$ gadmin backup restore <timepoint_to_restore> --time-point
----

For example, if you want to restore the cluster to the timepoint `2024-06-01T14:24:31Z` , run `gadmin backup restore 2024-06-01T14:24:31Z --time-point`. remember, the timepoint is precise to only seconds, other minor time parts will be truncated.

The point-in-time restore is based on full backup and differential backup. So we need to ensure the time point to restore is not under these cases:

> * the time points before the first full backup or after the current time.

Reviewer 2: ```suggestion
* Time Points Before the First Full Backup or After the Current Time:
- The selected time point cannot be earlier than the first full backup or later than the current system time. This is because the restore process needs a reference point from an existing full backup, and restoring to a future time is not possible.
- Example: If the first full backup was made on 2024-06-01T00:00:00Z, you cannot restore to any time before this, such as 2024-05-31T23:59:59Z. Similarly, if the current time is 2024-06-15T12:00:00Z, you cannot restore to a time like 2024-06-16T00:00:00Z.
```

----
$ gadmin backup restore <timepoint_to_restore> --time-point
----

For example, if you want to restore the cluster to the timepoint `2024-06-01T14:24:31Z` , run `gadmin backup restore 2024-06-01T14:24:31Z --time-point`. remember, the timepoint is precise to only seconds, other minor time parts will be truncated.

Point-in-Time Restore relies on both full and differential backups to restore the database to a specific time. When selecting a time point for restoration, it is crucial to ensure that the chosen time point meets certain conditions. Specifically, the time point must not fall under the following cases:

* Time Points Before the First Full Backup or After the Current Time:
- The selected time point cannot be earlier than the first full backup or later than the current system time. This is because the restore process needs a reference point from an existing full backup, and restoring to a future time is not possible.
- Example: If the first full backup was made on 2024-06-01T00:00:00Z, you cannot restore to any time before this, such as 2024-05-31T23:59:59Z. Similarly, if the current time is 2024-06-15T12:00:00Z, you cannot restore to a time like 2024-06-16T00:00:00Z.

> * the time points not covered by existing backups. Here "not covered" means the time points are not contained by any backup’s time points range [POINT-IN-TIME START TIME,  CREATED AT]. We can check the backup’s time points range by `gadmin backup list`. Here is an example:

Reviewer: ```suggestion
* Time Points Not Covered by Existing Backups
- The selected time point must be within the range of an existing backup’s coverage. A time point is considered "not covered" if it does not fall within the time range defined by any backup's POINT-IN-TIME START TIME and CREATED AT timestamps. You can verify the backup’s time ranges by running the gadmin backup list command. Here is an example:
```

====

== Key Benefits

The primary advantage of Point-in-Time Restore is the flexibility it offers, allowing users to restore the database to a specific time point. This feature is particularly useful in scenarios where unwanted changes or data corruption need to be undone without losing other valuable work done after the last backup.

.Key Benefit Scenario
====
. A user creates a full backup on __Sunday__.

. During the week, user creates a differential backup on every day 10pm.

. On Thursday, user finds there are unwanted changes made on Wednesday around 2pm.

. User can restore system to 1:55pm of Wednesday instead of the backup made on Tuesday 10pm.

. The changes made between Tuesday 10pm and Wednesday 1:55pm are saved this way.
====

Summery of key benefits:

* Precise recover time point to save a much as possible correct work made to database.

== Limitations and Considerations

Currently, the Point-in-Time Restore feature has these known limitations:

.Point-in-Time Restore examples.

> image::PIT_restore.png[]

Author: @victorleeTG  Here is the figure

= Point-in-Time Restore

> Point-in-Time Restore, introduced in version 4.1.0, enables users to restore the database to a previous time point using previously made backups, even though no backup was conducted at exactly that time point.

Reviewer 2: ```suggestion
Point-in-Time Restore, introduced in version 4.1, enables users to restore the database to a previous time point using previously made backups, even though no backup was conducted at exactly that time point.
```

> This feature is based on xref:tigergraph-server:backup-and-restore:differential-backups.adoc[Differential Backup].

Reviewer 2: ```suggestion
This feature replies on the user having made one or more xref:tigergraph-server:backup-and-restore:differential-backups.adoc[differential backups].
```

== Usage

To perform point-in-time restore, run the following command.
The timepoint should be in timestamp format.

[console, gsql]
----
$ gadmin backup restore <timepoint_to_restore> --time-point
----

> For example, if you want to restore the cluster to the timepoint `2024-06-01T14:24:31Z` , run `gadmin backup restore 2024-06-01T14:24:31Z --time-point`. remember, the timepoint is precise to only seconds, other minor time parts will be truncated.

Reviewer 2: ```suggestion
For example, if you want to restore the cluster to the time `2024-06-01T14:24:31Z` , run `gadmin backup restore 2024-06-01T14:24:31Z --time-point`

NOTE: The restore operation has 1 second granularity.  Fractional sections will be ignored.
```

| backup-2024-06-13T081939.120 | FULL        | 4.1.0   | 1.5 MB |                              | 2024-06-13T08:19:39Z     | 2024-06-13T08:19:39Z 
|================================================================================================================================================
---

In this example, the time range for backup `backup-2024-06-13T061939.131` is `[2024-06-13T06:19:39Z,2024-06-13T06:19:39Z]`, therefore, for a full backup is just a point. As for the incremental backup `backup-2024-06-13T063915.054`, its range is `[2024-06-13T06:19:39Z , 2024-06-13T06:39:15Z]`. `2024-06-13T06:35:00Z` is supported for it is covered by the time range `[2024-06-13T06:19:39Z , 2024-06-13T06:39:15Z]` of the backup `backup-2024-06-13T063915.054`. While, users can’t restore to the timepoint `2024-06-13T07:00:00Z` directly, for it is not in any backup's time range. However, if no restore was performed since `2024-06-13T07:00:00Z`, users may be able to create a new differential backup at right now and then restore to `2024-06-13T07:00:00Z`. 

[NOTE]
====

> because the restore action will make the uncovered timepoint unavailable to restore, if you are not very sure about the time point to restore,  we’d better ensure the possible time points are all covered by existing backups. if there is not this kind of backup, you may create a by `gadmin backup create <backup_tag> --incremental --base <full_backup_tag>`, the new backup will cover the range from the full backup <full_backup_tag>'s create time to the current time.

Reviewer 2: ```suggestion
because the restore action will make the uncovered timepoint unavailable to restore, if you are not very sure about the time point to restore,  it is best to ensure the possible time points are all covered by existing backups. If you do not yet have an appropriate backup, you may create one by running `gadmin backup create <backup_tag> --incremental --base <full_backup_tag>`, the new backup will cover the range from the full backup <full_backup_tag>'s create time to the current time.
```
